### DISCUSSION TITLE
Underscore custom extensions are not reset when creating new Language

### DISCUSSION THREAD

**Discussion by BramVanroy** [explosion/spaCy#5424] [Help: Coding & Implementations] [answered]
I found that one of my tests didn't pass a test as part of a suite, but it did pass when run in isolation. This was odd because the only difference between tests is that in the one that failed, I disabled a custom extension in a custom pipeline component. It seems (but I'm not sure) that if you have set an extension to Underscore once, this will always stay registered inside spaCy, even if you create a new Language instance. That is unexpected, I think.

Reproducible code:

```python
import spacy
from spacy.tokens import Doc


class CustomExt:
    name = 'test-component'

    def __init__(self, disable: bool = False):
        self.disable = disable
        self.set_extension()

    def set_extension(self):
        if not Doc.has_extension('test_ext') and not self.disable:
            Doc.set_extension('test_ext', default=None)

    def __call__(self, doc: Doc):
        # We need to hook the extensions again when using multiprocessing in Windows
        # see: https://github.com/explosion/spaCy/issues/4903
        # fixed in: https://github.com/explosion/spaCy/pull/5006
        self.set_extension()
        print(f"VALUE BEFORE SET w/ self.disable = {self.disable}: {doc._.test_ext}")
        if not self.disable:
            print("VALUE SET w/ self.disable = ", self.disable)
            doc._.set('test_ext', True)

        return doc


def init_parser(disable=False):
    nlp = spacy.load('en')
    test_comp = CustomExt(disable=disable)
    nlp.add_pipe(test_comp, last=True)
    return nlp


def main():
    nlp = init_parser(disable=False)
    print("Custom ext is disabled?", nlp.pipeline[-1][-1].disable)
    doc = nlp("hello world")
    print("VALUE AFTER SET w/ self.disable = False: ", doc._.test_ext)
    assert doc.has_extension("test_ext")

    print()
    
    # extension disabled
    nlp = init_parser(disable=True)
    print("Custom ext is disabled?", nlp.pipeline[-1][-1].disable)
    doc = nlp("hello world")
    print("VALUE AFTER SET w/ self.disable = True: ", doc._.test_ext)
    assert not doc.has_extension("test_ext")


if __name__ == '__main__':
    main()

```

OUTPUT
```
Custom ext is disabled? False
VALUE BEFORE SET w/ self.disable = False: None
VALUE SET w/ self.disable =  False
VALUE AFTER SET w/ self.disable = False:  True

Custom ext is disabled? True
VALUE BEFORE SET w/ self.disable = True: None
Traceback (most recent call last):
  File "dummy.py", line 54, in <module>
    main()
  File "dummy.py", line 50, in main
    assert not doc.has_extension("test_ext")
AssertionError
VALUE AFTER SET w/ self.disable = True:  None
```

We would expect that in the last case that a Doc does not get the custom `._.test_ext` property, since we passed `disable=True` to the custom component, effectively disabling any 'setting' of the attribute. However, in a newly initialized nlp object (accompanied by newly created custom component), these extensions are still set.

## Your Environment

Tested on both Windows and Linux:

* **spaCy version:** 2.2.4
* **Models:** en

on 

* **Platform:** Linux-4.15.0-46-generic-x86_64-with-debian-buster-sid
* **Python version:** 3.7.3

and

* **Platform:** Windows-10-10.0.18362-SP0
* **Python version:** 3.6.8

**Comment by adrianeboyd** [comment]
The extension state is global, so this is the expected behavior. From the docs (https://spacy.io/usage/processing-pipelines#custom-components-attributes):

> "Before you can access a custom extension, you need to register it using the set_extension method on the object you want to add it to, e.g. the Doc. Keep in mind that extensions are always added globally and not just on a particular instance."

**Comment by BramVanroy** [comment]
Any particular reason for that? It seems a bit unexpected from the user-side of things. It also makes it difficult to write tests for custom components. If you have any suggestions I'm happy to hear about how to deal with that.

Alternatively, is there a way to reset the Underscore?

**Comment by adrianeboyd** [comment, marked as answer]
I'm not sure if this is the whole story, but I think the global state allows the implementation to be simpler.

You can use `Doc/Span/Token.remove_extension` but I don't think there's a direct way to reset everything at once. This is how things are reset in the internal tests:

https://github.com/explosion/spaCy/blob/f333c2a0114589121b10ef93bd92eb27b7d78d2e/spacy/tests/doc/test_underscore.py#L10-L16

**Comment by BramVanroy** [comment]
> I'm not sure if this is the whole story, but I think the global state allows the implementation to be simpler.
> 
> You can use `Doc/Span/Token.remove_extension` but I don't think there's a direct way to reset everything at once. This is how things are reset in the internal tests:
> 
> https://github.com/explosion/spaCy/blob/f333c2a0114589121b10ef93bd92eb27b7d78d2e/spacy/tests/doc/test_underscore.py#L10-L16

That would work for my use-case. Perhaps some day I'll understand to how and why's of spaCy's internals. :) Thanks again!

**Comment by ines** [comment]
The deeper reasoning here is that the extension attributes are essentially mimicking native Python attributes, which are also global by default. So if you add an attribute to a class, you expect it to be present on every instance of it. Its behaviour may change based on the class state, but the attribute itself will always be present.

I remember we had a similar thread about this topic a quite a while ago (I think it was an issue) and there are definitely use cases where you want the attribute behaviour to change based on component state. But I'm not sure what the best and most elegant way to solve this is.
